pubsub/cloud-client/quickstart.py

```diff
@@ -19,76 +19,110 @@
     import time
 
+    from google.api_core.exceptions import NotFound
     from google.cloud import pubsub_v1
 
     # TODO project_id = "Your Google Cloud Project ID"
     # TODO topic_name = "Your Pub/Sub topic name"
+    # TODO subscription_name = "Your Pub/Sub subscription name"
     # TODO num_messages = number of messages to test end-to-end
 
-    # Instantiates a publisher and subscriber client
+    # Instantiate a publisher and a subscriber client
     publisher = pubsub_v1.PublisherClient()
     subscriber = pubsub_v1.SubscriberClient()
 
     # The `topic_path` method creates a fully qualified identifier
     # in the form `projects/{project_id}/topics/{topic_name}`
-    topic_path = subscriber.topic_path(project_id, topic_name)
+    topic_path = publisher.topic_path(project_id, topic_name)
 
     # The `subscription_path` method creates a fully qualified identifier
     # in the form `projects/{project_id}/subscriptions/{subscription_name}`
     subscription_path = subscriber.subscription_path(
         project_id, subscription_name)
 
-    # Create the topic.
-    topic = publisher.create_topic(topic_path)
-    print('\nTopic created: {}'.format(topic.name))
+    # Create a topic.
+    try:
+        publisher.delete_topic(topic_path)
+    except NotFound:
+        pass
+    finally:
+        topic = publisher.create_topic(topic_path)
+        print('Topic created: \"{}\"'.format(topic.name))
 
     # Create a subscription.
-    subscription = subscriber.create_subscription(
-        subscription_path, topic_path)
-    print('\nSubscription created: {}\n'.format(subscription.name))
-
-    publish_begin = time.time()
+    try:
+        subscriber.delete_subscription(subscription_path)
+    except NotFound:
+        pass
+    finally:
+        subscription = subscriber.create_subscription(
+            subscription_path, topic_path)
+        print('Subscription created: \"{}\"'.format(subscription.name))
+
+    # `data` is roughly 10 Kb.
+    data = 'x' * 9600
+    # `data` must be a bytestring.
+    data = data.encode('utf-8')
+    # Initialize an empty dictionary to track messages.
+    tracker = dict()
+    pubsub_time = 0.0
 
     # Publish messages.
-    for n in range(num_messages):
-        data = u'Message number {}'.format(n)
-        # Data must be a bytestring
-        data = data.encode('utf-8')
-        # When you publish a message, the client returns a future.
+    for i in range(num_messages):
+        # When we publish a message, the client returns a future.
         future = publisher.publish(topic_path, data=data)
-        print('Published {} of message ID {}.'.format(data, future.result()))
 
-    publish_time = time.time() - publish_begin
+        tracker.update({i: {'index': i,
+                            'pubtime': time.time(),
+                            'subtime': None}})
+
+        # `future.result()` blocks and returns a unique message ID per message.
+        tracker[future.result()] = tracker.pop(i)
```

Comment:
This seems wrong.  You should not have to block on this.  One unique identifier per message you have is the future itself.  So you can have a dictionary of tracker[publisher_future].  When that future returns, you could set it's message id in a callback or put it in a separate tracker keyed on the message id.

Reply:
When `pubtime` is captured, I feel it's captured just in time after `publish` is called. `tracker[future.result()] = tracker.pop(i)` switches in a new key for the info we already captured. Why do we need a separate tracker? 

```Python
    # Publish messages.
    for i in range(10):
        future = publisher.publish(topic_path, data=data)
        tracker.update({future: {'pubtime': time.time(),
                                 'subtime': None}})
        # Update the old key, which is the publish future, with the unique
        # identifier per message returned by the `result()` method.
        tracker[future.result()] = tracker.pop(future)
```

Reply:
I think what I was thinking here is that the call to future.result() blocks on the return of the publish and therefore all messages get published sequentially. This is neither good practice nor fast.  

The solution to this would be to do it in a non-blocking way.  Meaning,  I think, all updates to the tracker would have to be done in a callback for the future. 

Does that make any sense?  It could well be me misunderstanding your intent here.

Reply:
It makes sense now. Thank you!

This is the standard output after moving the code to resolve the future and to update the tracker into a callback: 

```
Published all messages.

Listening for messages...
7
5
1
6
3
8
9
0
2
4

Delivery Statistics
Average time: 2.336982s
Best time: 1.828190s
```